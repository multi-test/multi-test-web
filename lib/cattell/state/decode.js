import { BitStream } from 'bit-buffer';
import { fromBase64 } from "./base64";
function decodeAnswer(bitStream) {
    const bit2 = bitStream.readBits(2, false);
    switch (bit2) {
        case 0b01: return 'A';
        case 0b10: return 'B';
        case 0b11: return 'C';
        default: return '';
    }
}
function decodeGender(bitStream) {
    const bit2 = bitStream.readBits(2, false);
    switch (bit2) {
        case 0b01: return 'M';
        case 0b10: return 'F';
        default: return '';
    }
}
function decodeAge(bitStream) {
    const bit6 = bitStream.readBits(6, false);
    return bit6 === 0 ? 0 : 15 + bit6;
}
export function decodeState(base64) {
    const base64Data = base64.substring(0, 65);
    const bitStream = new BitStream(new ArrayBuffer(49));
    for (let i = 0; i < 65; i++) {
        const code = fromBase64(base64Data[i]);
        bitStream.writeBits(code, 6);
    }
    bitStream.index = 0;
    const position = bitStream.readUint8();
    const answers = new Array(187);
    for (let i = 0; i < 187; i++) {
        answers[i] = decodeAnswer(bitStream);
    }
    const gender = decodeGender(bitStream);
    const age = decodeAge(bitStream);
    const name = decodeURIComponent(escape(atob(base64.substring(65))));
    return {
        position,
        answers,
        profile: {
            gender,
            age,
            name,
        },
    };
}
