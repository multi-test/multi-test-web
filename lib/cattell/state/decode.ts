import {BitStream} from 'bit-buffer';
import {Blank, CattelAnswer, CattellState, Gender} from "./type";
import {decode as decodeKOI8U} from "../../utils/koi8-u";
import {byteLength, toByteArray} from "../../utils/base64";
import {crc16} from "../../utils/crc16";

function decodeAnswer(bitStream: BitStream): Blank<CattelAnswer> {
    const bit2 = bitStream.readBits(2, false);

    switch (bit2)  {
        case 0b01: return 'A';
        case 0b10: return 'B';
        case 0b11: return 'C';
        default: return '';
    }
}

function decodeGender(bitStream: BitStream): Blank<Gender> {
    const bit2 = bitStream.readBits(2, false);

    switch (bit2)  {
        case 0b01: return 'M';
        case 0b10: return 'F';
        default: return '';
    }
}

function decodeAge(bitStream: BitStream): number {
    const bit6 = bitStream.readBits(6, false);
    return bit6 === 0 ? 0 : 15 + bit6;
}

function convert64StringToBitStream(string64: string, sizeInBytes: number): BitStream {
    const bitStream = new BitStream(new ArrayBuffer(sizeInBytes));

    for (const byte of toByteArray(string64)) {
        bitStream.writeUint8(byte);
    }

    bitStream.index = 0;
    return bitStream;
}

function* readBitStream(bitStream: BitStream) {
    while (bitStream.bitsLeft > 16) {
        yield bitStream.readUint8();
    }
}

export function decodeState(string64: string): CattellState {
    const sizeInBytes = byteLength(string64);
    if (sizeInBytes < 51) {
        throw new Error('The provided Cattell test state is not complete');
    }
    const bitStream = convert64StringToBitStream(string64, sizeInBytes);
    const version = bitStream.readBits(2, false);
    if (version !== 0) {
        throw new Error(`Cannot parse an unsupported Cattell state serialization format (v${version})`);
    }

    const position = bitStream.readUint8();

    const answers = new Array<Blank<CattelAnswer>>(187);
    for (let i = 0; i < 187; i++) {
        answers[i] = decodeAnswer(bitStream);
    }

    const gender = decodeGender(bitStream);
    const age = decodeAge(bitStream);
    const name = decodeKOI8U(readBitStream(bitStream));
    const expectedChecksum = bitStream.readUint16();

    bitStream.index = 0;
    const actualChecksum = crc16(bitStream.readArrayBuffer(sizeInBytes - 2));

    if (actualChecksum !== expectedChecksum) {
        throw new Error(`Cattell state has been corrupted (actual = ${actualChecksum}, expected = ${expectedChecksum})`);
    }

    return {
        position,
        answers,
        profile: {
            gender,
            age,
            name,
        },
    };
}