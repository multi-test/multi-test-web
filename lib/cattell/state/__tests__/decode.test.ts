import {decodeState} from "../decode";

describe('cattell.state.decode', () => {
    let hash: string;

    beforeEach(() => {
        hash = "9HZ2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d2Z3d23enXwc7P1yDpLukuYpk=";
    });

    it('should decode hash to state ok', () => {
        expect(decodeState(hash)).toMatchSnapshot();
    });

    it('should forbid brute hacking into state', () => {
        hash = hash.substring(0, 10) + '0' + hash.substring(11);
        expect(() => decodeState(hash)).toThrowErrorMatchingSnapshot();
    });

    it('should throw error if protocol version is not 0', () => {
        hash = '9a' + hash.substring(2);
        expect(() => decodeState(hash)).toThrowErrorMatchingSnapshot();
    });

    it('should throw error if CRC16 is deleted from base state', () => {
        expect(() => decodeState("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA==")).toThrowErrorMatchingSnapshot();
    });
});

