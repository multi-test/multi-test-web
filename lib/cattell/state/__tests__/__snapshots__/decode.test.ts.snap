// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cattell.state.decode should decode hash to state ok 1`] = `
Object {
  "answers": Array [
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
    "C",
    "A",
    "B",
    "A",
    "C",
    "A",
  ],
  "position": 189,
  "profile": Object {
    "age": 70,
    "gender": "M",
    "name": "Иванов И.И.",
  },
}
`;

exports[`cattell.state.decode should forbid brute hacking into state 1`] = `"Cattell state has been corrupted (actual = 35360, expected = 39266)"`;

exports[`cattell.state.decode should throw error if CRC16 is deleted from base state 1`] = `"The provided Cattell test state is not complete"`;

exports[`cattell.state.decode should throw error if protocol version is not 0 1`] = `"Cannot parse an unsupported Cattell state serialization format (v1)"`;
